File: typescript/src/onboarding.ts
/**
 * The onboarding state machines, mirrored for optimistic UI.
 *
 * The authority is `gha-indie-worker-orm-core::enums` (backend only). This is
 * the same two machines, expressed so a wizard can advance immediately and
 * reconcile when the server answers — and so it can tell a *pending* step from
 * an *impossible* one. Keeping the edges here rather than in a component means
 * the web islands, Flutter and the desktop app cannot disagree about "next".
 */

export const ORG_ONBOARDING_STATES = [
  "created",
  "verifying_email",
  "awaiting_billing",
  "inviting_members",
  "connecting_repository",
  "first_run_pending",
  "active",
  "suspended",
] as const;
export type OrgOnboardingState = (typeof ORG_ONBOARDING_STATES)[number];

export const USER_ONBOARDING_STATES = [
  "registered",
  "verifying_email",
  "choosing_account_kind",
  "joining_org",
  "connecting_repository",
  "first_run_pending",
  "active",
  "disabled",
] as const;
export type UserOnboardingState = (typeof USER_ONBOARDING_STATES)[number];

const ORG_EDGES: Readonly<Record<OrgOnboardingState, readonly OrgOnboardingState[]>> = {
  created: ["verifying_email", "suspended"],
  verifying_email: ["awaiting_billing", "suspended"],
  awaiting_billing: ["inviting_members", "suspended"],
  inviting_members: ["connecting_repository", "suspended"],
  connecting_repository: ["first_run_pending", "suspended"],
  first_run_pending: ["active", "suspended"],
  active: ["suspended"],
  suspended: ["active"],
};

const USER_EDGES: Readonly<Record<UserOnboardingState, readonly UserOnboardingState[]>> = {
  registered: ["verifying_email", "disabled"],
  verifying_email: ["choosing_account_kind", "disabled"],
  choosing_account_kind: ["joining_org", "connecting_repository", "disabled"],
  joining_org: ["active", "connecting_repository", "disabled"],
  connecting_repository: ["first_run_pending", "disabled"],
  first_run_pending: ["active", "disabled"],
  active: ["disabled"],
  disabled: ["active"],
};

/** Percent-mille progress for a bar. `suspended` is zero: stopped, not nearly done. */
const ORG_PROGRESS: Readonly<Record<OrgOnboardingState, number>> = {
  created: 0,
  verifying_email: 167,
  awaiting_billing: 333,
  inviting_members: 500,
  connecting_repository: 667,
  first_run_pending: 833,
  active: 1000,
  suspended: 0,
};

export class TransitionError extends Error {
  readonly from: string;
  readonly to: string;

  constructor(machine: string, from: string, to: string) {
    super(`illegal ${machine} transition: ${from} -> ${to}`);
    this.name = "TransitionError";
    this.from = from;
    this.to = to;
  }
}

export function isOrgOnboardingState(value: string): value is OrgOnboardingState {
  return (ORG_ONBOARDING_STATES as readonly string[]).includes(value);
}

export function isUserOnboardingState(value: string): value is UserOnboardingState {
  return (USER_ONBOARDING_STATES as readonly string[]).includes(value);
}

export function orgSuccessors(state: OrgOnboardingState): readonly OrgOnboardingState[] {
  return ORG_EDGES[state];
}

export function userSuccessors(state: UserOnboardingState): readonly UserOnboardingState[] {
  return USER_EDGES[state];
}

/** The step a wizard's primary button advances to. Suspension is never it. */
export function orgNextStep(state: OrgOnboardingState): OrgOnboardingState | null {
  return ORG_EDGES[state].find((next) => next !== "suspended") ?? null;
}

export function orgTransition(
  from: OrgOnboardingState,
  to: OrgOnboardingState,
): OrgOnboardingState {
  if (!ORG_EDGES[from].includes(to)) {
    throw new TransitionError("org_onboarding", from, to);
  }
  return to;
}

export function userTransition(
  from: UserOnboardingState,
  to: UserOnboardingState,
): UserOnboardingState {
  if (!USER_EDGES[from].includes(to)) {
    throw new TransitionError("user_onboarding", from, to);
  }
  return to;
}

export function orgIsTerminal(state: OrgOnboardingState): boolean {
  return state === "active" || state === "suspended";
}

export function userIsTerminal(state: UserOnboardingState): boolean {
  return state === "active" || state === "disabled";
}

/**
 * `choosing_account_kind` is a fork — join an existing org, or connect your own
 * repository. A UI that shows one "next" button here picks for the user.
 */
export function userIsFork(state: UserOnboardingState): boolean {
  return state === "choosing_account_kind";
}

export function orgProgressPermille(state: OrgOnboardingState): number {
  return ORG_PROGRESS[state];
}

/**
 * An optimistic step. Immutable: `advance`, `reconcile` and `rollback` return a
 * new value rather than mutating, so a store can diff them.
 */
export interface Optimistic<S> {
  readonly confirmed: S;
  readonly pending: S | null;
}

export function optimistic<S>(confirmed: S): Optimistic<S> {
  return { confirmed, pending: null };
}

export function displayed<S>(value: Optimistic<S>): S {
  return value.pending ?? value.confirmed;
}

export function isPending<S>(value: Optimistic<S>): boolean {
  return value.pending !== null;
}

/** Move optimistically, but only along an edge the machine actually has. */
export function advanceOrg(
  value: Optimistic<OrgOnboardingState>,
  to: OrgOnboardingState,
): Optimistic<OrgOnboardingState> {
  orgTransition(displayed(value), to);
  return { confirmed: value.confirmed, pending: to };
}

export function advanceUser(
  value: Optimistic<UserOnboardingState>,
  to: UserOnboardingState,
): Optimistic<UserOnboardingState> {
  userTransition(displayed(value), to);
  return { confirmed: value.confirmed, pending: to };
}

/** The server's word wins, whatever was pending. */
export function reconcile<S>(_value: Optimistic<S>, server: S): Optimistic<S> {
  return { confirmed: server, pending: null };
}

export function rollback<S>(value: Optimistic<S>): Optimistic<S> {
  return { confirmed: value.confirmed, pending: null };
}

File: dart/lib/src/onboarding.dart
import 'errors.dart';

/// The onboarding state machines, mirrored for optimistic UI.
///
/// The authority is `gha-indie-worker-orm-core::enums` (backend only). Keeping
/// the edges here rather than in a widget means the Flutter app, the web
/// islands and the desktop app cannot disagree about what "next" is.
enum OrgOnboardingState {
  created('created'),
  verifyingEmail('verifying_email'),
  awaitingBilling('awaiting_billing'),
  invitingMembers('inviting_members'),
  connectingRepository('connecting_repository'),
  firstRunPending('first_run_pending'),
  active('active'),
  suspended('suspended');

  const OrgOnboardingState(this.wireName);

  final String wireName;

  static OrgOnboardingState? parse(String value) {
    for (final state in OrgOnboardingState.values) {
      if (state.wireName == value) {
        return state;
      }
    }
    return null;
  }

  List<OrgOnboardingState> get successors => switch (this) {
        OrgOnboardingState.created => const <OrgOnboardingState>[
            OrgOnboardingState.verifyingEmail,
            OrgOnboardingState.suspended,
          ],
        OrgOnboardingState.verifyingEmail => const <OrgOnboardingState>[
            OrgOnboardingState.awaitingBilling,
            OrgOnboardingState.suspended,
          ],
        OrgOnboardingState.awaitingBilling => const <OrgOnboardingState>[
            OrgOnboardingState.invitingMembers,
            OrgOnboardingState.suspended,
          ],
        OrgOnboardingState.invitingMembers => const <OrgOnboardingState>[
            OrgOnboardingState.connectingRepository,
            OrgOnboardingState.suspended,
          ],
        OrgOnboardingState.connectingRepository => const <OrgOnboardingState>[
            OrgOnboardingState.firstRunPending,
            OrgOnboardingState.suspended,
          ],
        OrgOnboardingState.firstRunPending => const <OrgOnboardingState>[
            OrgOnboardingState.active,
            OrgOnboardingState.suspended,
          ],
        OrgOnboardingState.active => const <OrgOnboardingState>[
            OrgOnboardingState.suspended,
          ],
        OrgOnboardingState.suspended => const <OrgOnboardingState>[
            OrgOnboardingState.active,
          ],
      };

  /// The step a wizard's primary button advances to. Suspension is reachable
  /// but is never the primary next step.
  OrgOnboardingState? get nextStep {
    for (final state in successors) {
      if (state != OrgOnboardingState.suspended) {
        return state;
      }
    }
    return null;
  }

  bool get isTerminal =>
      this == OrgOnboardingState.active || this == OrgOnboardingState.suspended;

  /// Per-mille progress for a bar. `suspended` is zero: stopped, not nearly done.
  int get progressPermille => switch (this) {
        OrgOnboardingState.created => 0,
        OrgOnboardingState.verifyingEmail => 167,
        OrgOnboardingState.awaitingBilling => 333,
        OrgOnboardingState.invitingMembers => 500,
        OrgOnboardingState.connectingRepository => 667,
        OrgOnboardingState.firstRunPending => 833,
        OrgOnboardingState.active => 1000,
        OrgOnboardingState.suspended => 0,
      };

  OrgOnboardingState transition(OrgOnboardingState to) {
    if (!successors.contains(to)) {
      throw IllegalTransition('org_onboarding', wireName, to.wireName);
    }
    return to;
  }
}

enum UserOnboardingState {
  registered('registered'),
  verifyingEmail('verifying_email'),
  choosingAccountKind('choosing_account_kind'),
  joiningOrg('joining_org'),
  connectingRepository('connecting_repository'),
  firstRunPending('first_run_pending'),
  active('active'),
  disabled('disabled');

  const UserOnboardingState(this.wireName);

  final String wireName;

  static UserOnboardingState? parse(String value) {
    for (final state in UserOnboardingState.values) {
      if (state.wireName == value) {
        return state;
      }
    }
    return null;
  }

  List<UserOnboardingState> get successors => switch (this) {
        UserOnboardingState.registered => const <UserOnboardingState>[
            UserOnboardingState.verifyingEmail,
            UserOnboardingState.disabled,
          ],
        UserOnboardingState.verifyingEmail => const <UserOnboardingState>[
            UserOnboardingState.choosingAccountKind,
            UserOnboardingState.disabled,
          ],
        UserOnboardingState.choosingAccountKind => const <UserOnboardingState>[
            UserOnboardingState.joiningOrg,
            UserOnboardingState.connectingRepository,
            UserOnboardingState.disabled,
          ],
        UserOnboardingState.joiningOrg => const <UserOnboardingState>[
            UserOnboardingState.active,
            UserOnboardingState.connectingRepository,
            UserOnboardingState.disabled,
          ],
        UserOnboardingState.connectingRepository => const <UserOnboardingState>[
            UserOnboardingState.firstRunPending,
            UserOnboardingState.disabled,
          ],
        UserOnboardingState.firstRunPending => const <UserOnboardingState>[
            UserOnboardingState.active,
            UserOnboardingState.disabled,
          ],
        UserOnboardingState.active => const <UserOnboardingState>[
            UserOnboardingState.disabled,
          ],
        UserOnboardingState.disabled => const <UserOnboardingState>[
            UserOnboardingState.active,
          ],
      };

  /// `choosingAccountKind` is a fork — join an existing org, or connect your
  /// own repository. A UI that shows one "next" button here picks for the user.
  bool get isFork => this == UserOnboardingState.choosingAccountKind;

  bool get isTerminal =>
      this == UserOnboardingState.active || this == UserOnboardingState.disabled;

  UserOnboardingState transition(UserOnboardingState to) {
    if (!successors.contains(to)) {
      throw IllegalTransition('user_onboarding', wireName, to.wireName);
    }
    return to;
  }
}

/// An optimistic step: the UI moves now and reconciles when the server answers.
/// Immutable, so a store can diff two values.
final class Optimistic<S> {
  const Optimistic(this.confirmed, [this.pending]);

  final S confirmed;
  final S? pending;

  S get displayed => pending ?? confirmed;
  bool get isPending => pending != null;

  /// The server's word wins, whatever was pending.
  Optimistic<S> reconcile(S server) => Optimistic<S>(server);

  Optimistic<S> rollback() => Optimistic<S>(confirmed);

  @override
  bool operator ==(Object other) =>
      other is Optimistic<S> &&
      other.confirmed == confirmed &&
      other.pending == pending;

  @override
  int get hashCode => Object.hash(confirmed, pending);
}

/// Move optimistically, but only along an edge the machine actually has.
Optimistic<OrgOnboardingState> advanceOrg(
  Optimistic<OrgOnboardingState> value,
  OrgOnboardingState to,
) {
  value.displayed.transition(to);
  return Optimistic<OrgOnboardingState>(value.confirmed, to);
}

Optimistic<UserOnboardingState> advanceUser(
  Optimistic<UserOnboardingState> value,
  UserOnboardingState to,
) {
  value.displayed.transition(to);
  return Optimistic<UserOnboardingState>(value.confirmed, to);
}
